src/Microsoft.AspNetCore.Razor.Language/DefaultRazorIntermediateNodeLoweringPhase.cs

```diff
@@ -673,22 +821,229 @@ public override void VisitTagHelperBlock(TagHelperChunkGenerator chunkGenerator,
 
                 _builder.Push(new TagHelperBodyIntermediateNode());
 
-                VisitDefault(block);
+                VisitList(node.Body);
 
                 _builder.Pop(); // Pop InitializeTagHelperStructureIntermediateNode
 
-                AddTagHelperAttributes(tagHelperBlock.Attributes, tagHelperBlock.Binding);
+                Visit(node.StartTag);
 
                 _builder.Pop(); // Pop TagHelperIntermediateNode
+
+                // No need to visit the end tag because we don't write any IR for it.
+
+                // We don't want to track attributes from a previous tag helper element.
+                _renderedBoundAttributeNames.Clear();
```

Comment:
Kind of jank to maintain state here. I get why you have to but just calling it out if you can think of a better way.